### ISSUE TITLE
Mono on Trusty, legacy infrastructure - fails to install

### ISSUE THREAD

**Issue by jen20** [travis-ci/travis-ci#4974]
Not sure if this is because of the Xamarin packaging infrastructure, but I'm seeing the following logs for either Mono 3.12.1 or latest, when running in the legacy environment (with `sudo: required` enabled):

```
Using worker: travis-worker-gce-org-prod-2:836d3f98-85e2-496a-b9f7-b03f15ce664a (testing-gce-89f7187a-f0a5-4b8f-a22d-46d61fe6cafb:travis-ci-mega-trusty-1444853315)
system_info
Build system information
Build language: csharp
Build dist: trusty
Build image provisioning date and time
Wed Oct 14 20:47:30 UTC 2015
Operating System Details
Distributor ID:	Ubuntu
Description:	Ubuntu 14.04.3 LTS
Release:	14.04
Codename:	trusty
Linux Version
3.19.0-28-generic
Cookbooks Version
fffffff https://github.com/travis-ci/travis-cookbooks/tree/fffffff
Git version
git version 1.9.1
GCC version
gcc (Ubuntu 4.8.4-2ubuntu1~14.04) 4.8.4
Copyright (C) 2013 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
LLVM version
clang version 3.5.0 (tags/RELEASE_350/final)
Target: x86_64-unknown-linux-gnu
Thread model: posix
Pre-installed Ruby versions
jruby-9.0.1.0
ruby-1.9.3-p551
ruby-2.0.0-p647
ruby-2.1.7
ruby-2.2.3
Pre-installed Node.js versions
iojs-v1.6
iojs-v1.6.4
v0.10
v0.10.40
v0.11.16
v0.12.2
v0.12.7
v0.6.21
v0.8.28
v4.1.2
Pre-installed Go versions
1.0.3
1.1.2
1.2.2
1.3.3
1.4.2
1.5.1
Pre-installed PostgreSQL versions
9.1.19
9.2.14
9.3.10
9.4.5
Redis version
redis-server 3.0.4
RabbitMQ Version
3.5.4
Installed Sphinx versions
2.0.10
2.1.9
2.2.8
Default Sphinx version
2.2.8
Installed Firefox version
firefox 31.6.0esr
ant -version
Apache Ant(TM) version 1.9.3 compiled on April 8 2014
mvn -version
Apache Maven 3.1.1 (0728685237757ffbf44136acec0402957f723d9a; 2013-09-17 15:22:22+0000)
Maven home: /usr/local/maven
Java version: 1.8.0_60, vendor: Oracle Corporation
Java home: /usr/lib/jvm/java-8-oracle/jre
Default locale: en_US, platform encoding: UTF-8
OS name: "linux", version: "3.19.0-28-generic", arch: "amd64", family: "unix"
gradle -version
------------------------------------------------------------
Gradle 2.7
------------------------------------------------------------
Build time:   2015-09-14 07:26:16 UTC
Build number: none
Revision:     c41505168da69fb0650f4e31c9e01b50ffc97893
Groovy:       2.3.10
Ant:          Apache Ant(TM) version 1.9.3 compiled on December 23 2013
JVM:          1.8.0_60 (Oracle Corporation 25.60-b23)
OS:           Linux 3.19.0-28-generic amd64
lein version
Downloading Leiningen to /home/travis/.lein/self-installs/home/travis/.lein/leiningen-2.5.1-standalone.jar now...
Leiningen 2.5.1 on Java 1.8.0_60 Java HotSpot(TM) 64-Bit Server VM
phpenv versions
* system (set by /home/travis/.phpenv/version)
C# support for Travis-CI is community maintained.
Please open any issues at https://github.com/travis-ci/travis-ci/issues/new and cc @joshua-anderson @akoeplinger @nterry
Installing Mono
Executing: gpg --ignore-time-conflict --no-options --no-default-keyring --homedir /tmp/tmp.aGEdnWbhMF --no-auto-check-trustdb --trust-model always --keyring /etc/apt/trusted.gpg --primary-keyring /etc/apt/trusted.gpg --keyring /etc/apt/trusted.gpg.d/apt.postgresql.org.gpg --keyserver keyserver.ubuntu.com --recv-keys 3FA7E0328081BFF6A14DA29AA6A19B38D3D831EF
gpg: requesting key D3D831EF from hkp server keyserver.ubuntu.com
gpg: key D3D831EF: public key "Xamarin Public Jenkins (auto-signing) <releng@xamarin.com>" imported
gpg: Total number processed: 1
gpg:               imported: 1  (RSA: 1)
6.61sW: Failed to fetch http://download.mono-project.com/repo/debian/dists/wheezy/snapshots/3.12.1/main/binary-amd64/Packages  404  Not Found [IP: 54.230.88.227 80]
E: Some index files failed to download. They have been ignored, or old ones used instead.
The command "sudo apt-get update -qq" failed and exited with 100 during .
Your build has been stopped.
```

My .travis.yml file looks like this:

```yaml
language: csharp
sudo: required
dist: trusty
mono:
  - latest
  - 3.12.1
script:
- ./build.sh 0.0.0.0
- ./run_tc_tests.sh
```

Is there something I'm missing here? /cc (as directed): @joshua-anderson @akoeplinger @nterry

**Comment by solarce** [comment]
Howdy @jen20, so you have `dist: trusty` in your file, which means the builds will run under our new [Trusty beta](http://docs.travis-ci.com/user/trusty-ci-environment/), not the Legacy precise based infrastructure.

It's looking like some changes to https://github.com/travis-ci/travis-build/blob/master/lib/travis/build/script/csharp.rb may be needed to best support Trustry

**Comment by jen20** [comment]
Thanks for the information - is the best solution here to revert back to running on precise in the short term?

**Comment by akoeplinger** [comment]
Hey James, we already got a report in https://github.com/travis-ci/travis-ci/issues/4907 and I implemented the necessary changes in https://github.com/travis-ci/travis-build/pull/529.

We're just waiting on a deploy from Travis, @meatballhat said he'd do it today but I guess it got delayed a bit due the infrastructure issues over the weekend :)

**Comment by jen20** [comment]
Thanks @akoeplinger - in that case I'll wait for the deployment and give this another shot. Since this is already tracked elsewhere I'll follow that issue and close this one.

**Comment by BanzaiMan** [comment]
@akoeplinger The PR has been merged and deployed to staging. Could you test it?

**Comment by akoeplinger** [comment]
@BanzaiMan I already tested it on staging a few days ago, it should be good to go :)

**Comment by BanzaiMan** [comment]
@akoeplinger Awesome. I've deployed this to production. Thanks!

**Comment by jen20** [comment]
:+1: all seems good now. Worth nothing that even if you explicitly want Mono 3.12.1, you need to specify 3.12.0, probably because of Xamarin package repository oddities.

**Comment by akoeplinger** [comment]
@jen20 http://docs.travis-ci.com/user/languages/csharp/#Choosing-Mono-version-to-test-against

> Note: even if you specify e.g. 3.12.0 the version used by your build may actually be 3.12.1 depending on what the latest version in the 3.12.x series is (it’s a limitation of the Xamarin repositories right now).

**Comment by jen20** [comment]
Yes, I was aware of that - it seems that even if you explicitly *only* want 3.12.1 (e.g. because that's the version you will statically link against) you have to specify 3.12.0 and rely on that behaviour though (i.e. the opposite of the warning).

**Comment by akoeplinger** [comment]
@jen20 yep, that's correct. The gist is that you can only use versions that correspond to a folder in the APT repo here: http://download.mono-project.com/repo/debian/dists/wheezy/snapshots/. What's contained in there is another story :)
